### ISSUE TITLE
belos: Evaluate test coverage 

### ISSUE THREAD

**Issue by tlamonthezie** [NexGenAnalytics/Trilinos#55]
Before splitting into particular issues per package, survey the current state of tests in belos package that have only Epetra unit tests (and not a Tpetra equivalent).

Evaluate the number of Tpetra tests that would be needed to add to ensure comparable code test coverage.

Relates to #50

**Comment by tlamonthezie** [comment]
The tests of belos package are organized in sub directories.
There are epetra/test and an tpetra/test directories but it does not always follow the same naming.
The **Missing** information indicate that there might be a tpetra correspondance but we have to analyze the code implementation because the file names do not match

| Test subfolder                                          | Epetra Tests (Main file name)                                  | Requirements              | Tpetra Tests                                                    | Requirements                           | Correspondance in Tpetra (if sure) |
|---------------------------------------------------------|----------------------------------------------------------------|---------------------------|-----------------------------------------------------------------|----------------------------------------|------------------------------------|
| BiCGStab                                                | BiCGStabExFile.cpp                                             | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| BiCGStab                                                | PrecBiCGStabExFile.cpp                                         | Epetra+Ifpack             | ---                                                             | ---                                    | **Missing**                        |
| BiCGStab                                                | test_bicgstab_diag.cpp                                         | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| BiCGStab                                                | ---                                                            | ---                       | test_bicgstab_hb.cpp                                            | Tpetra+Triutils                        | *Tpetra Only*                      |
| BlockCG                                                 | test_bl_cg_hb.cpp                                              | Epetra+Triutils           | test_bl_cg_hb.cpp                                               | Tpetra                                 | **OK**                             |
| BlockCG                                                 | test_bl_pcg_hb.cpp                                             | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | test_bl_pcg_set_res.cpp                                        | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | test_pseudo_cg_hb.cpp                                          | Epetra+Triutils           | test_pseudo_bl_cg_hb.cpp                                                            | Tpetra                                    | **OK**                        |
| BlockCG                                                 | test_pseudo_cg_indefinite.cpp                                  | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | test_pseudo_pcg_hb.cpp                                         | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | test_pseudo_stochastic_pcg_hb.cpp                              | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | test_resolve_cg_hb.cpp                                         | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| BlockCG                                                 | ---                                                            | ---                       | test_bl_cg_hb_multiprec.cpp                                     | **NOT REGISTERED IN CMakeLists.txt ?** | *Tpetra Only*                      |
| BlockCG                                                 | ---                                                            | ---                       | test_pseudo_stochastic_cg_hb.cpp                                | Tpetra                                 | *Tpetra Only*                      |
| BlockCG                                                 | ---                                                            | ---                       | test_bl_cg_complex_hb.cpp                                       | Tpetra+Triutils                        | *Tpetra Only*                      |
| BlockGmres                                              | test_bl_fgmres_hb.cpp                                          | Epetra+Triutils+Ifpack    | test_bl_fgmres_hb.cpp                                           | Tpetra                                 | **OK**                             |
| BlockGmres                                              | test_bl_gmres_hb.cpp                                           | Epetra+Triutils           | test_bl_gmres_hb.cpp                                            | Tpetra+Triutils                        | **OK**                             |
| BlockGmres                                              | test_pseudo_gmres_hb.cpp                                       | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| BlockGmres                                              | test_resolve_gmres_hb.cpp                                      | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| BlockGmres                                              | test_bl_gmres_diag.cpp                                         | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| BlockGmres                                              | test_bl_pgmres_hb.cpp                                          | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| BlockGmres                                              | test_pseudo_pgmres_hb.cpp                                      | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| BlockGmres                                              | ---                                                            | ---                       | test_bl_gmres_hb_df.cpp                                         | Tpetra+Triutils                        | *Tpetra Only*                      |
| BlockGmres                                              | ---                                                            | ---                       | test_hybrid_gmres_hb.cpp                                        | Tpetra                                 | *Tpetra Only*                      |
| Epetra64Driver                                          | cxx_main.cpp                                                   | Epetra                    | ---                                                             | ---                                    | *Tpetra Only*                      |
| FixedPoint                                              | test_fp_hb.cpp                                                 | Epetra+Triutils           | test_fp_hb.cpp                                                  | Tpetra                                 | **OK**                             |
| GCRODR                                                  | test_gcrodr_hb.cpp                                             | Epetra+Triutils           | test_gcrodr_hb.cpp                                              | Tpetra+Triutils                        | **OK**                             |
| GCRODR                                                  | test_prec_gcrodr_hb.cpp                                        | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| GCRODR                                                  | ---                                                            | ---                       | test_block_gcrodr.cpp                                           | Tpetra+Triutils+Experimental           | *Tpetra Only*                      |
| GCRODR                                                  | ---                                                            | ---                       | test_gcrodr_complex_hb.cpp                                      | Tpetra+Triutils                        | *Tpetra Only*                      |
| hypre                                                   | hypre_UnitTest.cpp                                             | Epetra+Ifpack+Galeri      | ---                                                             | ---                                    | **Missing**                        |
| Tpetra/LinearSolverFactory (**=Epetra SolverFactory?**) | ---                                                            | ---                       | LinearSolverFactory.cpp                                         | Tpetra                                 | *Tpetra Only*                      |
| Tpetra/LinearSolverFactory (**=Epetra SolverFactory?**) | ---                                                            | ---                       | CustomSolverFactory.cpp                                         | Tpetra                                 | *Tpetra Only*                      |
| Tpetra/LinearSolverFactory (**=Epetra SolverFactory?**) | ---                                                            | ---                       | SolverFactory.cpp  (**equiv(Epetra test_solver_factory.cpp)?**) | Tpetra                                 | *Tpetra Only*                      |
| SolverFactory (**=Tpetra LinearSolverFactory?**)        | test_solver_factory.cpp (**equiv(Tpetra SolverFactory.cpp)?**) | Epetra+Galeri             | ---                                                             | ---                                    | **Missing**                        |
| MINRES                                                  | test_minres_diag.cpp                                           | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| MINRES                                                  | test_minres_hb.cpp                                             | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| MINRES                                                  | test_minres_indefinite.cpp                                     | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| MINRES                                                  | test_pminres_hb.cpp                                            | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| MultiMatrixSolve                                        | test_bl_gmres_multidiag.cpp                                    | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| MultiMatrixSolve                                        | test_pseudo_gmres_multi_hb.cpp                                 | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| MultipleSolve (**=Epetra MultiMatrixSolve?**)           | ---                                                            | ---                       | Tpetra_MultipleSolves.cpp                                       | Tpetra                                 | *Tpetra Only*                      |
| MVOPTester                                              | cxx_main.cpp                                                   | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| MVOPTester                                              | ---                                                            | ---                       | cxx_main_complex.cpp                                            | Tpetra                                 | *Tpetra Only*                      |
| Native                                                  | ---                                                            | ---                       | diagonal_matrix.cpp                                             | Tpetra                                 | *Tpetra Only*                      |
| Native                                                  | ---                                                            | ---                       | nonsymm_matrix.cpp                                              | Tpetra                                 | *Tpetra Only*                      |
| OrthoManager                                            | belos_orthomanager_epetra.cpp                                  | Epetra+Triutils           | belos_orthomanager_tpetra.cpp                                   | Tpetra+Triutils                        | **OK**                             |
| PCPG                                                    | test_pcpg_epetraex.cpp                                         | Epetra+Triutils+EpetraExt |                                                                 |                                        | **Missing**                        |
| RCG                                                     | test_prec_rcg_hb.cpp                                           | Epetra+Triutils+EpetraExt | ---                                                             | ---                                    | **Missing**                        |
| RCG                                                     | test_rcg_hb.cpp                                                | Epetra+Triutils+EpetraExt | test_rcg_hb.cpp                                                 | Tpetra+Triutils                        | **OK**                             |
| TFQMR                                                   | test_pseudo_ptfqmr_hb.cpp                                      | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| TFQMR                                                   | test_pseudo_tfqmr_hb.cpp                                       | Epetra+Triutils           | ---                                                             | ---                                    | **Missing**                        |
| TFQMR                                                   | test_ptfqmr_hb.cpp                                             | Epetra+Triutils+Ifpack    | ---                                                             | ---                                    | **Missing**                        |
| TFQMR                                                   | test_tfqmr_diag.cpp                                            | Epetra                    | ---                                                             | ---                                    | **Missing**                        |
| TFQMR                                                   | test_tfqmr_hb.cpp                                              | Epetra+Triutils           | test_tfqmr_hb.cpp                                               | Tpetra+Triutils                        | **OK**                             |

**Comment by stmcgovern** [comment]
@tlamonthezie @antoinemeyer5  The MinRes and PCPG are highest priority in the Belos package. Please add the missing Tpetra equivalents before focusing on other tests in this package.

**Comment by stmcgovern** [comment]
I have split the PCPG test addition into another issue. Please continue to work on the MINRES test.

**Comment by jennloe** [comment]
Hi all, 
I've been talking with @hkthorn to clarify exactly which tests need to be converted to Tpetra. 
For now, I can say with certainty that 
1. Epetra64Driver does NOT need to be converted; that was Epetra-specific. 
2. All tests ending in _diag and _indefinite definitely MUST be converted. 
3. Test requiring Ifpack need to be put on hold for the time being pending resolution of the dependency issue. 
4. We also need to convert the files in the "Examples" directory, as these are used as human-readable tests.

Hopefully that gives you more information to run with for now. I'll have more details soon about correspondences between some of the less-obvious tpetra vs epetra tests.

**Comment by stmcgovern** [comment]
thanks @jennloe

**Comment by jennloe** [comment]
More comments on test conversion:

BlockCG solve: 
1. for test_bl_cg_hb.cpp: Add a test with the "--use-single-red" option as in the CMakeLists for Epetra tests. 
2. (I think the 2nd row for BlockCG in your table is extra?)
3. test_pseudo_cg_hb.cpp in Epetra is equivalent to test_pseudo_bl_cg_hb.cpp in Tpetra.
4. test_bl_pcg_set_res.cpp DOES require ifpack. (Should be updated in CMakeLists if not shown.)
5. You know that test_pseudo_cg_indefinite.cpp needs to be converted. (Maybe there is already a PR.)
6. test_resolve_cg_hb.cpp Needs conversion as well. 
7. test_pseudo_stochastic_pcg_hb.cpp in Epetra is almost equivalent to test_pseudo_stochastic_cg_hb.cpp in Tpetra; the latter just lacks preconditioning. We are still working to resolve the ifpack2 dependency issue, so wait on this test.

**Comment by jennloe** [comment]
BlockGmres:
1. `test_bl_gmres_hb.cpp` Tpetra version needs a "--pseudo" option and test as in Epetra version. 
2. `test_pseudo_gmres_hb.cpp` `test_resolve_gmres_hb.cpp` and `test_bl_gmres_diag.cpp` all need to be converted. 

For the others, we are still waiting on the Ifpack question.

BiCGStab: 
1. `BiCGStabExFile.cpp` is equivalent to Tpetra `test_bicgstab_hb.cpp`, so no conversion needed. 
2. `test_bicgstab_diag.cpp` needs conversion
3. Waiting on ifpack question for the last one.

**Comment by jennloe** [comment]
- Don't worry about tests in the `MVOPTester`, `FixedPoint`,`OrthoManager`, or `Native` directories. These are fine as they are.
